ksql-engine/src/main/java/io/confluent/ksql/query/QueryExecutor.java

```diff
@@ -182,6 +182,63 @@ public TransientQueryMetadata buildTransientQuery(
     );
   }
 
+  public QueryMetadata buildTransientQuery(
+      final String statementText,
+      final QueryId queryId,
+      final Set<SourceName> sources,
+      final ExecutionStep<?> physicalPlan,
+      final String planSummary,
+      final LogicalSchema schema,
+      final RowConsumer rowConsumer
+  ) {
+    final KsqlQueryBuilder ksqlQueryBuilder = queryBuilder(queryId);
+    final PlanBuilder planBuilder = new KSPlanBuilder(ksqlQueryBuilder);
+    final Object buildResult = physicalPlan.build(planBuilder);
+    final KStream<?, GenericRow> kstream;
+    if (buildResult instanceof KStreamHolder<?>) {
+      kstream = ((KStreamHolder<?>) buildResult).getStream();
+    } else if (buildResult instanceof KTableHolder<?>) {
+      final KTable<?, GenericRow> ktable = ((KTableHolder<?>) buildResult).getTable();
+      kstream = ktable.toStream();
+    } else {
+      throw new IllegalStateException("Unexpected type built from exection plan");
+    }
+
+    kstream.foreach((k, row) -> {
+      if (row == null) {
+        return;
+      }
+      rowConsumer.accept(row);
+    });
```

Comment:
this seems a little too hacked together for my liking - it's like making a modification to the physical/logical plan "outside" of the physical and logical planners. Not sure I have suggestions at the moment, but give it a thought

Reply:
It's pretty much an exact copy and paste of the existing code for transient queries

Reply:
the main difference is that this adds a `forEach`, which actually affects the execution plan/topology

Reply:
The foreach is from here https://github.com/confluentinc/ksql/blob/master/ksql-engine/src/main/java/io/confluent/ksql/query/TransientQueryQueue.java#L59

I've basically just simplified the pre-existing code into a single place. There's no new logic though.